Database-Features/removing-corrupt-data.md

```diff
@@ -0,0 +1,91 @@
+
+Save the below script into a file and run the below script using [Exaplus CLI](https://docs.exasol.com/db/latest/connect_exasol/sql_clients/exaplus_cli/exaplus_cli.htm). It will generate a list of SQL statements which forces the database to read all columns of every table in the entire database. 
+
+```sql
+set heading off;
+set verbose off;
+set linesize 20000;
+spool /tmp/check_all_columns.sql;
+SELECT '/*"' || COLUMN_SCHEMA || '"."' || COLUMN_TABLE || '"*/ ' || 'SELECT '||GROUP_CONCAT(CASE WHEN column_type='BOOLEAN' THEN 'COUNT("'||column_name||'")' ELSE CASE WHEN column_type LIKE '%CHAR%' THEN 'MIN(LENGTH("'||column_name||'"))' ELSE 'MIN("'||column_name||'")' END END ORDER BY column_ordinal_position)||' FROM "'||column_schema||'"."'||column_table||'";' FROM SYS."$EXA_SYS_COLUMNS_BASE" where column_schema='EXA_STATISTICS' and column_object_type='TABLE' group by column_table,column_schema order by column_table desc;
+SELECT '/*"' || COLUMN_SCHEMA || '"."' || COLUMN_TABLE || '"*/ ' || 'SELECT '||GROUP_CONCAT(CASE WHEN column_type='BOOLEAN' THEN 'COUNT("'||column_name||'")' ELSE CASE WHEN column_type LIKE '%CHAR%' THEN 'MIN(LENGTH("'||column_name||'"))' ELSE 'MIN("'||column_name||'")' END END ORDER BY column_ordinal_position)||' FROM "'||column_schema||'"."'||column_table||'";' FROM SYS.EXA_DBA_COLUMNS where column_object_type='TABLE' group by column_table,column_schema order by column_table desc;
+spool /tmp/check_all_columns.log;
+set heading on;
+set verbose on;
+@/tmp/check_all_columns.sql;
+```
+To execute the file, run `exaplus -u sys -p <passwd> -c <connection string> -f <file name>`. 
+
+The script will run through the entire script and store the output into /tmp/check_all_column.log. Once it's finished, do the following:
+1. Look into the output and find any queries which return an error message. The table which causes an error message contains corrupt data and should be dropped.
+2. Save the DDL of the object. Many SQL Editors offer the ability to save the DDL of the object, however we also have a script to help with this. For more information, see [Create DDL for a Table](create-ddl-for-a-table.md).
+3. Drop the table. For more information, see [DROP TABLE](https://docs.exasol.com/db/latest/sql/drop_table.htm).
+4. Re-create the table using the saved DDL.
+5. Reload the data into the table from another source. 
+
+If one of the statistics tables fails, contact Exasol to delete the corresponding data. Once all of the corrupt objects are dropped, re-run the above script to make sure there are no errors. 
+
+If the database is still crashing after all of the tables are fixed, you can try to drop all of the indexes in case one of them was also corrupted. The steps for this are:
+
+1. (optional) Run the below query to generate ENFORCE INDEX statements.
+```sql
```

Comment:
I believe we should use the same SQL script approach as in the detection script. This here comes a bit "quick & dirty". There might be plenty of indices.

Reply:
I would also remove the optional. Dropped indices should be recreated manually in any case - at least we have to recommend this. There have been too many support cases where dropped indices caused massive system performance outages.

Reply:
@MichaelFleischer do you mean to add the DROP INDEX/ENFORCE INDEX statements into the original SQL script? As a consequence, we would say that we ALWAYS just drop the indexes, select all of the columns/tables, and then re-create the indexes. This may cause the script to be very long if the indexes also need to be re-created and this step could be saved.

Reply:
No, I did not mean that. 
My point was that we have 3 sections or scripts that generate SQL. But the quality of SQL script we create differs. The first one creates a plug-and-play convenient version but the others do not. It is not clear why not to provide the same service to all three.

Reply:
I put the index SQL script in the same format as the other one, so someone only needs to run one command in exaplus. Please see the new script and double check my syntax!